### DISCUSSION TITLE
Out of range exception when adding new values to C-style arrays

### DISCUSSION THREAD

**Discussion by NotpainRaov** [nlohmann/json#4256] [Q&A]
Hi everyone. I am having issues with parsing C-style arrays using the library. When adding a new element to the array, then trying to read data from a file which was last updated before the new value was added to the array (so the data we read and parse has less values than expected), this exception gets thrown: 

> [json.exception.out_of_range.401] array index 5 is out of range

The array (saved as a json list) in the file on disk contains 5 values, but after modifying the array in the code, it's supposed to have 6.
Here is the example code used:
```
bool TestBoolArr[6] = { true, true, true, true, true, true };
```

And that part of the file on disk looks like this:
```
"TestBoolArr": [
            false,
            false,
            false,
            true,
            true
],
```
        
Are there any workarounds for this? Switching the raw array to an std::vector works, but I would prefer not doing so if it's not necessary.

**Comment by gregmarr** [comment]
You will have to do the extraction from JSON manually in the object containing the array, rather than using any of the predefined functions that read the array based on the current size of the array.  If you share your current code, we may be able to help you with that.

**Reply by NotpainRaov** [reply]
Thanks for the reply. Here is the current code:
```cpp
class ClassSerializationTest
{
public:
    bool    TestBoolArr[6] = { true, true, true, true, true, true };

    NLOHMANN_DEFINE_TYPE_INTRUSIVE(ClassSerializationTest, TestBoolArr)
};
```

Parsing/loading from file:
```cpp
std::ifstream inputs(*filePath, std::ios::binary);
if (inputs.fail())
{
    return;
}

std::string fileData((std::istreambuf_iterator<char>(inputs)), std::istreambuf_iterator<char>());
nlohmann::json parsedData = nlohmann::json::parse(fileData);

inputs.close();

testClassInstance = parsedData.get<ClassSerializationTest>();
```

**Reply by gregmarr** [reply]
You'll need to do something like the following.  I've verified that it compiles using Compiler Explorer, but haven't run it.

```
class ClassSerializationTest
{
public:
    bool    TestBoolArr[6] = { true, true, true, true, true, true };

    // Custom from_json function that supports reading the values from a shorter array and leaving the other values at their current values.
    friend void from_json(const nlohmann::json& j, ClassSerializationTest& t) { 
      auto &data = j.at("TestBoolArr");
      auto size = std::min(data.size(), std::size_t(6));
      for (std::size_t i = 0; i < size; ++i)
      {
        TestBoolArr[i] = data.at(i).get<bool>();
      }
    }

    // Can continue to use the standard to_json function.
    NLOHMANN_DEFINE_TYPE_INTRUSIVE_ONLY_SERIALIZE(ClassSerializationTest, TestBoolArr)
};
```
